content/program-testing.md

```diff
@@ -0,0 +1,488 @@
+# Program Testing
+
+### List links for further reading here:
+
+- [Solana Cookbook](https://github.com/ixmorrow/solana-program-unit-tests.git)
+- [Rust Book - Testing](https://doc.rust-lang.org/book/ch11-00-testing.html)
+
+# Lesson Objectives
+
+*By the end of this lesson, you will be able to:*
+
+- Understand all of the different ways how to test Solana programs
+- Know the difference between unit tests, integration tests, and RPC tests
+- Debug Solana programs
+
+# TL;DR
+
+- Testing is a **key component** of smart contract development, it ensures the code works as intended before releasing it to the public.
+- Solana programs support three different types of tests (unit, integration, and client-side) all of which have a specific purpose.
+
+# Overview
+
+## Testing Concepts
+
+Testing in software is very common and there are actually entire careers fields dedicated to just creating and running tests. A robust testing process can minimize the amount of bugs developers introduce into production code by catching them before they pose a real issue. You obviously cannot test *everything*, but it’s important to try to think of all the ways you can try to break your program or cause some unintended actions with your tests. This is especially imperative when developing smart contracts because a single bug can lead to millions of dollars lost or stolen.
+
+### Tests in Rust
+
+The rust package manager, Cargo, natively has some tools built into it to help developers write their own automated tests. Whenever we make a new library project with Cargo, a test module with a test function in it is automatically generated for us.
+
+The Rust community thinks about tests in terms of two main categories: unit tests and integration tests. *Unit tests* are small and more focused, testing one module in isolation at a time, and can test private interfaces. *Integration tests* are entirely external to your library and use your code in the same way any other external code would, using only the public interface and potentially exercising multiple modules per test.
+
+The purpose of unit tests is to test each unit of code in isolation from the rest of the code to quickly pinpoint where code is and isn’t working as expected. Unit tests reside in the `src` directory in the file with the code they are testing. Unit tests are declared inside a module named `tests` annotated with `cfg(test)`. At its simplest, a test in Rust is a function that’s annotated with the `#[test]` attribute.
```

Comment:
This section came almost directly from the rust book https://doc.rust-lang.org/stable/book/ch11-03-test-organization.html so that's what I was using as reference

Reply:
@jamesrp13 is this resolved?

Reply:
I just reworded it slightly